### ISSUE TITLE
Discovering All User Identities (GET users/discover)

### ISSUE THREAD

**Issue by leogdion** [brightdigit/MistKit#28]

**Comment by leogdion** [comment]
## Auth + routing constraints discovered while attempting to integration-test `users/discover`

While verifying #308 (v1.0.0-beta.1 prep) the MistDemo integration suite exposed two CloudKit Web Services constraints that affect both the GET (this issue) and POST (#27) variants:

### 1. The endpoint is rejected on the private database
Calling the POST variant against `private` returns:
```
HTTP 400 BAD_REQUEST
"BadRequestException: endpoint not applicable in the database type 'privatedb'"
```
Verified live against `iCloud.com.brightdigit.MistDemo`. So the request must route through `.../public/users/discover` regardless of which database the calling app primarily uses. The same is implied for the GET variant by Apple's docs ([DiscoveringAllUserIdentities](https://developer.apple.com/library/archive/documentation/DataManagement/Conceptual/CloudKitWebServicesReference/DiscoveringAllUserIdentities.html#//apple_ref/doc/uid/TP40015240-CH31-SW1)).

### 2. The endpoint requires user-context auth (web-auth token)
Calling either variant from a public-DB service signed with **server-to-server** credentials returns `HTTP 401 AUTHENTICATION_FAILED`. So the service has to be configured with:
- `database: .public`
- `tokenManager: WebAuthTokenManager(apiToken:webAuthToken:)`

### Implication for MistDemo and `MistKitClientFactory`
The current `MistKitClientFactory` / `DatabaseCredentials` convention pairs `.public → S2S` and `.private/.shared → web-auth`. Neither produces the `.public + web-auth` combination this endpoint needs. Three approaches to consider:

1. **Auxiliary service in `PhaseContext`** — let user-identity phases optionally use a second `CloudKitService` configured for public+web-auth. Confined to MistDemo's integration runner; doesn't change the library.
2. **Database-credentials union** — extend `DatabaseCredentials` to allow `.publicWebAuth(apiToken:webAuthToken:)`, surface it to `MistKitClientFactory.create(for:)`. Wider impact, but lets external CLIs target `users/discover` without bespoke wiring.
3. **Document the limitation only** — leave the integration test gap as known, note in CLAUDE.md that this endpoint requires the unusual combo.

### Status
The POST variant (#27) shipped in v1.0.0-alpha.5 and is closed, but it has never been exercised by the MistDemo integration suite for the reasons above — `DiscoverUserIdentitiesPhase` was wired into `PrivateDatabaseTest`, which is why the suite hit constraint #1. As of #309 the phase has been removed from `PrivateDatabaseTest.phases` (with a comment explaining why) and is currently unreferenced; whatever approach this issue lands on for GET should bring back coverage for both variants.

cc #27 (POST variant, closed but untested via the integration runner)

**Comment by leogdion** [comment]
## Follow-up: library state and a small improvement

Verified the library side after the auth/routing analysis above. **MistKit already supports `WebAuthTokenManager + .public` today** — there's no library gate on that combination.

### Current `CloudKitService` initializers (`Sources/MistKit/Service/CloudKitService+Initialization.swift`)

| Initializer | `database` |
|---|---|
| `init(containerIdentifier:apiToken:webAuthToken:transport:)` | hardcoded `.private` (line 52) |
| `init(containerIdentifier:apiToken:transport:)` | hardcoded `.public` (line 77, comment "API-only supports public database") |
| `init(containerIdentifier:tokenManager:environment:database:transport:)` | **`Database = .private` parameter — accepts any value** |

The URLSession convenience copies (lines 129, 146, 161) mirror the same shape. So the path that already works for public+web-auth is:

```swift
let tm = WebAuthTokenManager(apiToken: apiToken, webAuthToken: webAuthToken)
let service = try CloudKitService(
    containerIdentifier: containerIdentifier,
    tokenManager: tm,
    environment: .development,
    database: .public
)
```

This is exactly the shape `MistDemoKit`'s `DatabaseCredentials.makeTokenManager()` already produces for `.privateDatabase`/`.sharedDatabase` after #285. The only thing missing on the MistDemo side is a `.publicWebAuth(apiToken:webAuthToken:)` variant on `DatabaseCredentials`.

### Suggested library-level cleanup (small, separate from #28's GET implementation)

Both convenience initializers should accept an optional `database:` parameter (default keeps current behavior). That removes the implicit "API-only ⇒ public" / "web-auth ⇒ private" assumption baked into the convenience layer:

```swift
public init(
  containerIdentifier: String,
  apiToken: String,
  webAuthToken: String,
  database: Database = .private,   // new — currently hardcoded
  transport: any ClientTransport
) throws

public init(
  containerIdentifier: String,
  apiToken: String,
  database: Database = .public,    // new — currently hardcoded
  transport: any ClientTransport
) throws
```

Source-compatible for callers that don't pass `database:`. Worth filing as its own issue if not folded into the #28 work.

### Updated path forward for this issue

When implementing GET `users/discover` (and re-enabling integration coverage for POST per #27):
1. Add `.publicWebAuth(apiToken:webAuthToken:)` variant to `DatabaseCredentials` in MistDemoKit.
2. Add an aux `userContextService: CloudKitService?` to `PhaseContext` (built from that variant when web-auth credentials are available).
3. Optionally also add the `database:` parameter to the two convenience inits above, so the library surface matches what's already true generically.

**Comment by leogdion** [comment]
## Other endpoints with the same public+web-auth requirement

The user-identity discovery group in CloudKit Web Services all shares this routing/auth combination. Apple's reference shows them under `public/users/...` paths and they all require web-auth credentials (e.g. `users/lookup/email` example uses `public/users/lookup/email`). The same `DatabaseCredentials.publicWebAuth` + `userContextService` plumbing this issue lands should also unblock:

- **#33** — Fetching Contacts (`users/lookup/contacts`)
- **#34** — Fetching Users by Email (`users/lookup/email`) — Apple's curl example explicitly uses the `public/` path
- **#35** — Fetching Users by Record Name (`users/lookup/id`)

Plus #27 (POST `users/discover`, already implemented but untested via the integration suite for the same reason).

### Adjacent endpoints — **not** in this category

- `users/current` / `users/caller`: documented as `public/users/current` in Apple's reference, but our integration test ran it successfully on `private` + web-auth, so the routing is more lenient here than for `users/discover`. No new variant needed; works with the existing private+web-auth convention.
- `records/accept` (#42) and other sharing endpoints: these route through `.shared`, not `.public`. The existing `MistKitClientFactory` convention (private/shared → web-auth) already supports them.
- `tokens/create` (#52) / `tokens/register` (#53): APNs push token registration — typically API-token only, no user context. Outside this gap.

### Recommendation

Whoever picks up the GET implementation here should treat the user-discovery + user-lookup family (#28, #27, #33, #34, #35) as one slice — they share the same `DatabaseCredentials.publicWebAuth` requirement on the MistDemo side and (optionally) the same `database:` parameterization improvement on the two convenience inits noted in the previous comment.

**Comment by leogdion** [comment]
## Implementation handoff (for picking up in a fresh session)

Consolidating the previous three comments into a single actionable plan. All file paths are relative to repo root unless noted.

### TL;DR
CloudKit's user-identity endpoints (`users/discover` GET+POST, `users/lookup/contacts`, `users/lookup/email`, `users/lookup/id`) require **public-database routing + web-auth credentials**. MistDemo's current `MistKitClientFactory` convention pairs `.public → S2S` and `.private/.shared → web-auth`, so there's no path that produces public + web-auth — the combination CloudKit needs. The library itself already supports this combination (verified — see "Library state" comment); only MistDemo's `DatabaseCredentials` enum lacks the variant.

### Background context
- Discovered while running `mistdemo test-private --verbose` against the live container during PR #309 (v1.0.0-beta.1 prep).
- POST `users/discover` was rejected on `private` with `BAD_REQUEST: endpoint not applicable in the database type 'privatedb'`.
- POST against `public` + S2S returned `AUTHENTICATION_FAILED` 401 — confirming user-context (web-auth) is required.
- `DiscoverUserIdentitiesPhase` was removed from `PrivateDatabaseTest.phases` in PR #309 commit `02cae41`. The phase source file currently lives at `Examples/MistDemo/Sources/MistDemoKit/Integration/Phases/DiscoverUserIdentitiesPhase.swift` and is referenced by no test pipeline — bringing it back is part of this work.

### Files to change

**MistDemoKit (required):**
1. `Examples/MistDemo/Sources/MistDemoKit/Configuration/DatabaseCredentials.swift`
   - Add `case publicWebAuth(apiToken: String, webAuthToken: String)` to `DatabaseCredentials`.
   - `database` accessor returns `.public` for it.
   - `makeTokenManager()` returns `WebAuthTokenManager(apiToken:webAuthToken:)`.
   - Update `MistDemoConfig.toDatabaseCredentials()` to construct this variant when web-auth credentials are present and the integration test asks for it (likely via a new flag or a separate factory entry point — see #285 for the bundled-credentials pattern).

2. `Examples/MistDemo/Sources/MistDemoKit/Integration/PhaseContext.swift`
   - Add `internal let userContextService: CloudKitService?` (defaulting to `nil` in the existing memberwise init).

3. `Examples/MistDemo/Sources/MistDemoKit/Integration/IntegrationTestRunner.swift`
   - Add an optional `userContextService: CloudKitService?` parameter; thread into `makeContext()`.

4. `Examples/MistDemo/Sources/MistDemoKit/Integration/Phases/FetchCurrentUserPhase.swift`
   - Prefer `context.userContextService ?? context.service` for `fetchCurrentUser()` — keeps private-DB behavior, lets public-DB tests opt in.

5. `Examples/MistDemo/Sources/MistDemoKit/Integration/Phases/DiscoverUserIdentitiesPhase.swift`
   - Require `context.userContextService` (skip with a clear message when nil — public+web-auth credentials not configured).

6. `Examples/MistDemo/Sources/MistDemoKit/Integration/Tests/PublicDatabaseTest.swift`
   - Append `FetchCurrentUserPhase()` then `DiscoverUserIdentitiesPhase()` to `phases`.

7. `Examples/MistDemo/Sources/MistDemoKit/Commands/TestIntegrationCommand.swift`
   - Build the `userContextService` from `MistDemoConfig` (using `.publicWebAuth` credentials if available) and pass to the runner.

8. `Examples/MistDemo/Sources/MistDemoKit/CloudKit/MistKitClientFactory.swift`
   - Add a helper like `createUserContextService(for:) throws -> CloudKitService?` that returns `nil` when the user-context credentials aren't configured.

**MistKit library (optional cleanup, can be a separate PR — see "Library state" comment):**
9. `Sources/MistKit/Service/CloudKitService+Initialization.swift`
   - Add `database: Database = .private` parameter to the `apiToken+webAuthToken+transport` init (currently hardcodes `.private` at line 52).
   - Add `database: Database = .public` parameter to the `apiToken+transport` init (currently hardcodes `.public` at line 77).
   - Mirror in the URLSession convenience copies (lines 129, 146).
   - Source-compatible — defaults preserve existing behavior.

### Implementation phases / endpoint coverage

The same plumbing unblocks five issues — file separately or fold together:
- **#28 (this issue)** — GET `users/discover`
- **#27** (closed but untested) — POST `users/discover`
- **#33** — `users/lookup/contacts`
- **#34** — `users/lookup/email`
- **#35** — `users/lookup/id`

### Verification

After implementation, set both credential sets in `Examples/MistDemo/.env` (the user already has them populated for the private-DB test):
```
CLOUDKIT_KEY_ID=...               # for the existing public+S2S test
CLOUDKIT_PRIVATE_KEY_PATH=...     # for the existing public+S2S test
CLOUDKIT_API_TOKEN=...            # for the new public+web-auth aux service
CLOUDKIT_WEB_AUTH_TOKEN=...       # for the new public+web-auth aux service
```
Then:
```bash
cd Examples/MistDemo
set -a; source .env; set +a
swift run mistdemo test-integration --verbose
```
Expected: the existing 8 phases pass with S2S, then the two new user-identity phases pass against the public DB with web-auth.

### Provenance
- PR #309 (`308-v100-beta1` branch) — established the `DatabaseCredentials` enum and removed `DiscoverUserIdentitiesPhase` from `PrivateDatabaseTest`. Use that PR's commit `02cae41` as the reference point for the integration-test wiring.
- Findings here came from running `mistdemo test-private --verbose` and `mistdemo test-integration --verbose` against `iCloud.com.brightdigit.MistDemo` on 2026-05-08.

**Comment by leogdion** [comment]
## Live verification finding (2026-05-08)

While landing #312, all the OpenAPI / Swift plumbing for this endpoint shipped (path builder, response processor, generated client, public method) but **live testing returned HTTP 500** from Apple. The other user-identity endpoints in the same family work — \`users/caller\`, POST \`users/discover\`, \`users/lookup/email\`, and \`users/lookup/id\` all responded successfully against \`iCloud.com.brightdigit.MistDemo\` in the same run.

**Run output (relevant excerpt):**

\`\`\`
👤 Fetch caller (current user)
✅ Caller: _904181d1d76652d1f53581aec11cf76e

👥 Discover user identities
✅ Discovered 1 user identity

🌐 Discover all user identities
2026-05-08T14:56:25-0400 warning com.brightdigit.MistKit.api: Unhandled response status code: 500
❌ Error: httpError(statusCode: 500)
\`\`\`

**Hypothesis on why it 500s.** The \"GET \`/users/discover\`\" form is referenced as a navigation breadcrumb in Apple's archived CloudKit Web Services Reference, but I can't find an actual REST endpoint definition for it. CloudKitJS exposes \`Container.discoverAllUserIdentities()\` as a method, but the JS SDK may translate it to a different REST shape (perhaps POST with empty body, or a different path entirely). The 500 is consistent with \"endpoint exists at the gateway but the dispatch is misrouted.\"

**Status in PR #315.** Per [memory rule](https://github.com/brightdigit/MistKit/issues/29#issuecomment-...) about capturing follow-ups in issues:
- \`CloudKitService.discoverAllUserIdentities()\` is shipped but marked \`@available(*, unavailable, message: \"...\" )\` pointing at this issue.
- The MistDemo \`DiscoverAllUserIdentitiesPhase\` and the unit-test directory have been removed.
- The OpenAPI entry, generated client, path builder, response processor, and Swift wrapper are all retained — unblocking will be a one-line \`@available\` removal once we figure out the correct REST shape.

**Suggested next steps for whoever picks this up:**
1. Probe with curl using the same web-auth credentials and try variations:
   - POST \`/users/discover\` with empty body \`{}\` (vs the current POST shape that takes \`lookupInfos\`)
   - GET \`/users/discover\` with a query string version of \`lookupInfos\`
   - POST \`/users/discover\` with \`lookupInfos: []\`
2. Compare against CloudKitJS's outgoing request when calling \`discoverAllUserIdentities()\` (DevTools network panel against the CloudKit JS demo).
3. Check whether the endpoint requires a discoverable account in the caller's Address Book — the test container may have an empty contact list, which could trigger a 500 instead of an empty array.

**Comment by leogdion** [comment]
## Update — 2026-05-08: Definitive Apple-side defect (REST + CloudKitJS both broken)

Deep investigation on branch [`28-delete-after-apple-feedback-filed`](https://github.com/brightdigit/MistKit/tree/28-delete-after-apple-feedback-filed) confirms this is an **Apple-side server defect**, reproducible through both the REST endpoint and Apple's own CloudKitJS library. MistKit's `discoverAllUserIdentities()` stays `@available(*, unavailable)`. The OpenAPI definition, generated client, path builder, response processor, and Swift wrapper all remain in place — unblocking is a one-line `@available` removal once Apple fixes the handler.

### What we proved

1. **The REST URL we send matches Apple's documented spec exactly.** `GET /database/1/{container}/development/public/users/discover` per the [archived REST docs](https://developer.apple.com/library/archive/documentation/DataManagement/Conceptual/CloudKitWebServicesReference/DiscoveringAllUserIdentities.html).

2. **CloudKitJS sends the byte-for-byte equivalent request.** Disassembled `cdn.apple-cloudkit.com/ck/2/cloudkit.js`:
   ```js
   discoverAllUserIdentities: function() {
     var e = (new RequestBuilder())
       .setApiEntityName("users")
       .setApiActionName("discover")
       .setResponseClass(UserIdentitiesResponse);
     return this.publicCloudDatabase.sendRequest(e);
   }
   // builder default: this._method = isNullOrUndefined(c) ? "GET" : "POST"
   // setPayload() flips to POST. discoverAllUserIdentities never calls setPayload.
   ```
   Same builder pattern as the working `fetchCurrentUserIdentity` → `users/caller` (which confirms our REST `fetchCaller()` works fine).

3. **REST probe matrix proves the URL is reachable, recognized, and authenticated** ([probe script in tree](https://github.com/brightdigit/MistKit/blob/28-delete-after-apple-feedback-filed/Examples/MistDemo/Scripts/probes/discover-get-500.sh)):

   | Probe | Status | Note |
   |---|---|---|
   | `OPTIONS users/discover` | **200 OK** with `Access-Control-Allow-Methods: GET, POST, PUT, DELETE` | Apple advertises GET as allowed |
   | `GET users/discoverAll` (typo) | **404 NOT_FOUND** with `"could not find handler for endpoint 'GET discoverAll'"` | Unknown paths return NOT_FOUND, not 500 |
   | `GET private/users/discover` | **400** `"endpoint not applicable in the database type 'privatedb'"` | Confirms `/public/` segment is required |
   | `POST users/discover` (control) | **400** `"discover request contains an invalid user id"` | Same URL + auth, handler runs, validates body |
   | `GET users/discover` × 3 variants | **500 INTERNAL_ERROR** | Handler crashes |

   Every 500 response includes a refreshed `X-Apple-CloudKit-Web-Auth-Token` header — auth is succeeding, then the handler throws.

4. **CloudKitJS canary confirms the same defect inside Apple's own library.** Added a "Test discoverAllUserIdentities" button to `mistdemo auth-token`'s sign-in page that calls `container.discoverAllUserIdentities()` directly. Result from a fully-authenticated browser session, 2026-05-08 19:43:20 UTC:

   ```json
   {
     "status": "rejected",
     "CKError.uuid": "26a433f0-d9df-483a-b086-9b35b527ce8a",
     "CKError.serverErrorCode": "INTERNAL_ERROR",
     "CKError.ckErrorCode": "INTERNAL_ERROR",
     "CKError.reason": "INTERNAL_ERROR"
   }
   ```

   ~331ms round-trip. Same defect, observed via Apple's own library.

5. **Native CloudKit framework context.** `CKDiscoverAllUserIdentitiesOperation` itself has no top-level "Deprecated" banner, but: it carries both a deprecated `discoverAllUserIdentitiesCompletionBlock` and a newer `Result`-based variant; the page's "Deprecated classes" cross-reference lists `CKDiscoverUserIdentitiesOperation` (sibling) as deprecated; and the whole flow depends on `requestApplicationPermission(.userDiscoverability)` — exactly the contacts-correlation surface Apple has been quietly dismantling for privacy reasons. The 500 (server crash, not 404 or 410) suggests the server-side handler has been retired without formally deprecating the API surface.

### Apple correlation UUIDs

For Feedback Assistant or any iCloud team escalation. Container: `iCloud.com.brightdigit.MistDemo`, environment: `development`.

| Source       | Method | UUID                                     | Time (UTC)             |
| ------------ | ------ | ---------------------------------------- | ---------------------- |
| REST probe 2 | GET    | \`fd63e695-214f-4ad7-90d0-f78966809285\` | 2026-05-08T19:23:59Z   |
| REST probe 3 | GET    | \`6c09078c-71d8-46a1-907d-bcb5604b6306\` | 2026-05-08T19:23:59Z   |
| REST probe 4 | GET    | \`deb4efc6-fc08-41b3-a2e1-fe6fa6585395\` | 2026-05-08T19:24:00Z   |
| CloudKitJS   | GET    | \`26a433f0-d9df-483a-b086-9b35b527ce8a\` | 2026-05-08T19:43:20Z   |

### Artifacts on the branch

- [`Examples/MistDemo/Scripts/probes/discover-get-500.sh`](https://github.com/brightdigit/MistKit/blob/28-delete-after-apple-feedback-filed/Examples/MistDemo/Scripts/probes/discover-get-500.sh) — runnable REST probe matrix
- [`Examples/MistDemo/Scripts/probes/discover-get-500-findings.md`](https://github.com/brightdigit/MistKit/blob/28-delete-after-apple-feedback-filed/Examples/MistDemo/Scripts/probes/discover-get-500-findings.md) — full evidence chain
- [`Examples/MistDemo/Scripts/probes/apple-feedback-draft.md`](https://github.com/brightdigit/MistKit/blob/28-delete-after-apple-feedback-filed/Examples/MistDemo/Scripts/probes/apple-feedback-draft.md) — paste-in body for Apple Feedback Assistant
- `AuthTokenIndexHTML+ScriptDiscoverTest.swift` — permanent canary button on the auth-token sign-in page; will surface the moment Apple fixes the endpoint

### Reproduction

\`\`\`bash
# REST side
./Examples/MistDemo/Scripts/probes/discover-get-500.sh

# CloudKitJS side
swift run mistdemo auth-token --no-browser --port 8082
# Open http://127.0.0.1:8082, sign in, click "Test discoverAllUserIdentities".
\`\`\`

**Comment by leogdion** [comment]
# Apple Feedback Assistant — Draft

Use this as a paste-in body for a Feedback Assistant report.

---

**Area / category:** CloudKit → CloudKit Web Services / CloudKit JS

**Title:** `discoverAllUserIdentities` returns HTTP 500 INTERNAL_ERROR via REST and CloudKitJS

**Description:**

Apple's `discoverAllUserIdentities` endpoint is broken on the server. It is reproducibly returning `HTTP 500` with `serverErrorCode: "INTERNAL_ERROR"` from both the documented REST endpoint and from Apple's own `cloudkit.js` library. The failure happens after authentication succeeds (auth tokens are correctly refreshed in the response headers), so it is a server-side handler crash rather than an auth or routing problem.

**Steps to reproduce:**

REST path:

```
GET https://api.apple-cloudkit.com/database/1/{container}/development/public/users/discover
    ?ckAPIToken={api-token}
    &ckWebAuthToken={character-map-encoded-web-auth-token}
```

CloudKitJS path: in a browser authenticated to a CloudKit container, call

```js
container.discoverAllUserIdentities()
```

**Expected:** `200 OK` with a `users` array (per the [archived REST docs][archive] and the [CloudKitJS docs][cloudkitjs]).

**Actual:** `500 Internal Server Error` with body:

```json
{
  "uuid": "<correlation-uuid>",
  "serverErrorCode": "INTERNAL_ERROR"
}
```

CloudKitJS surfaces the same failure as a `CKError { serverErrorCode: "INTERNAL_ERROR" }` rejection.

**Diagnostic evidence (rules out client-side wire-shape issues):**

1. `OPTIONS` on the same URL returns `200 OK` with `Access-Control-Allow-Methods: GET, POST, PUT, DELETE` — the GET method is registered on this URL.
2. A typo'd path on the same prefix (e.g. `/users/discoverAll`) returns `404 NOT_FOUND` with `"could not find handler for endpoint 'GET <name>'"` — so unknown paths return `NOT_FOUND`, not `INTERNAL_ERROR`. Our path is recognized.
3. Sending `GET /users/discover` against `private/` or `shared/` databases returns `400 BAD_REQUEST` with `"endpoint not applicable in the database type 'privatedb'/'sharedb'"` — confirming the public-database segment is correct.
4. `POST /users/discover` against the identical URL with the identical auth credentials reaches the handler and runs body validation (returns `400 BAD_REQUEST: discover request contains an invalid user id` for an empty `lookupInfos` array).
5. Auth succeeds even on the failing GET requests: every 500 response includes a refreshed `X-Apple-CloudKit-Web-Auth-Token` header, which CloudKit only emits after authentication is accepted.
6. Apple's own CloudKitJS (served from `cdn.apple-cloudkit.com/ck/2/cloudkit.js`) sends the byte-for-byte equivalent request — the request builder defaults to `GET` when no payload is set, and `discoverAllUserIdentities` never calls `setPayload`. Calling `container.discoverAllUserIdentities()` from an authenticated browser produces the same 500.

**Affected scope:** Both the REST API and CloudKitJS. Sibling endpoints on the same path prefix (`users/caller`, `POST users/discover`) work normally with the same auth and host.

**Apple correlation UUIDs from the affected container `iCloud.com.brightdigit.MistDemo`, environment `development`:**

| Source       | Method | UUID                                     | Time (UTC)             |
| ------------ | ------ | ---------------------------------------- | ---------------------- |
| REST         | GET    | `fd63e695-214f-4ad7-90d0-f78966809285`   | 2026-05-08T19:23:59Z   |
| REST         | GET    | `6c09078c-71d8-46a1-907d-bcb5604b6306`   | 2026-05-08T19:23:59Z   |
| REST         | GET    | `deb4efc6-fc08-41b3-a2e1-fe6fa6585395`   | 2026-05-08T19:24:00Z   |
| CloudKitJS   | GET    | `26a433f0-d9df-483a-b086-9b35b527ce8a`   | 2026-05-08T19:43:20Z   |

**Frequency:** 100% reproducible.

**Suggested resolution:** If the endpoint has been retired, return a documented status (`410 GONE` or `400` with an explicit `serverErrorCode` such as `ENDPOINT_RETIRED`) and update the [archived REST docs][archive] / [CloudKitJS docs][cloudkitjs] accordingly. Returning `INTERNAL_ERROR` to a request with the documented shape is misleading and prevents downstream code from handling the deprecation cleanly.

**External tracking:** [brightdigit/MistKit#28](https://github.com/brightdigit/MistKit/issues/28).

[archive]: https://developer.apple.com/library/archive/documentation/DataManagement/Conceptual/CloudKitWebServicesReference/DiscoveringAllUserIdentities.html
[cloudkitjs]: https://developer.apple.com/documentation/cloudkitjs/cloudkit.container/discoveralluseridentities

**Comment by leogdion** [comment]
## Preserved investigation artifacts

The branch [`28-delete-after-apple-feedback-filed`](https://github.com/brightdigit/MistKit/tree/28-delete-after-apple-feedback-filed) (commit `1cd612d`) is being closed without merging — only the unique investigation commit lived there, and its evidence has been folded into this issue. The two pieces of source that would otherwise be lost (probe script + CloudKitJS canary) are inlined verbatim below so they remain runnable even after the branch is deleted.

---

### 1. REST probe script — `discover-get-500.sh`

Original path on the branch: `Examples/MistDemo/Scripts/probes/discover-get-500.sh`. Reads credentials from `Examples/MistDemo/.env` (gitignored). Run from `Examples/MistDemo/Scripts/probes/`.

```bash
#!/usr/bin/env bash
#
# Probe script for issue #28: Apple's GET /users/discover returns
# HTTP 500 INTERNAL_ERROR. This script exists as documentation of the
# investigation in branch `28-delete-after-apple-feedback-filed`.
#
# What the probes prove:
#   - GET on /database/1/<container>/development/public/users/discover
#     returns 500 INTERNAL_ERROR consistently, with no body other than
#     `{ uuid, serverErrorCode: "INTERNAL_ERROR" }`.
#   - Apple's OPTIONS preflight on the same URL returns 200 with
#     `Access-Control-Allow-Methods: GET, POST, PUT, DELETE`, so the
#     endpoint *is* registered server-side.
#   - Routing is correct: a typo path (e.g. /users/discoverAll) returns
#     404 with `serverErrorCode: NOT_FOUND, reason: "could not find
#     handler for endpoint 'GET discoverAll'"`. We don't get that.
#   - Auth is correct: every 500 response includes a refreshed
#     `X-Apple-CloudKit-Web-Auth-Token`, which CloudKit only emits after
#     successful authentication.
#   - Database routing is correct: GET on /private/users/discover and
#     /shared/users/discover return 400 with explicit
#     "endpoint not applicable in the database type 'privatedb'/'sharedb'".
#
# Conclusion: the URL we send matches Apple's documented spec
# (see https://developer.apple.com/library/archive/documentation/DataManagement/Conceptual/CloudKitWebServicesReference/DiscoveringAllUserIdentities.html)
# but Apple's handler crashes on the GET. Server-side defect.
#
# Reads credentials from `Examples/MistDemo/.env` (gitignored).

set -u

cd "$(dirname "$0")/../.."

if [ ! -f .env ]; then
  echo "Missing Examples/MistDemo/.env" >&2
  exit 1
fi

set -a
# shellcheck disable=SC1091
source ./.env
set +a

API_TOKEN="${CLOUDKIT_API_TOKEN:?missing}"
WEB_AUTH_TOKEN="${CLOUDKIT_WEB_AUTH_TOKEN:?missing}"
CONTAINER="${CLOUDKIT_CONTAINER_IDENTIFIER:?missing}"

# CloudKit's character map for ckWebAuthToken: + -> %2B, / -> %2F, = -> %3D
ENCODED_WEB_AUTH=$(printf '%s' "$WEB_AUTH_TOKEN" | sed -e 's/+/%2B/g' -e 's|/|%2F|g' -e 's/=/%3D/g')
QUERY="ckAPIToken=${API_TOKEN}&ckWebAuthToken=${ENCODED_WEB_AUTH}"

probe() {
  local label="$1"; local method="$2"; local url="$3"; shift 3
  echo "================================================================"
  echo "${label}"
  echo "${method} ${url}"
  echo "================================================================"
  curl -sS -i -X "${method}" "$@" "${url}?${QUERY}" 2>&1 \
    | grep -E '^(HTTP/|Server:|Date:|X-Apple-Request-UUID:|Allow:|Access-Control-Allow-Methods:|\{|  ")' \
    | head -20
  echo
}

BASE_DEV_PUBLIC="https://api.apple-cloudkit.com/database/1/${CONTAINER}/development/public/users/discover"

probe "1. POST with empty array (control: confirms POST handler reachable)" \
  POST "${BASE_DEV_PUBLIC}" \
  -H 'Content-Type: application/json' -d '{"lookupInfos": []}'

probe "2. GET (no body, no extra headers) — the failure" \
  GET "${BASE_DEV_PUBLIC}"

probe "3. GET with Accept: application/json" \
  GET "${BASE_DEV_PUBLIC}" \
  -H 'Accept: application/json'

probe "4. GET with explicit empty body" \
  GET "${BASE_DEV_PUBLIC}" \
  --data-binary ''

probe "5. GET on private DB (control: 400 'endpoint not applicable')" \
  GET "https://api.apple-cloudkit.com/database/1/${CONTAINER}/development/private/users/discover"

probe "6. GET on shared DB (control: 400 'endpoint not applicable')" \
  GET "https://api.apple-cloudkit.com/database/1/${CONTAINER}/development/shared/users/discover"

probe "7. GET fake path /users/discoverAll (control: 404 NOT_FOUND with handler-name reason)" \
  GET "https://api.apple-cloudkit.com/database/1/${CONTAINER}/development/public/users/discoverAll"

probe "8. OPTIONS — confirms GET is in Allow-Methods (this is the smoking gun)" \
  OPTIONS "${BASE_DEV_PUBLIC}"

echo "================================================================"
echo "Summary:"
echo "  Probe 8 (OPTIONS) confirms Apple advertises GET as allowed."
echo "  Probe 7 (fake path) confirms unknown paths return NOT_FOUND, not 500."
echo "  Probe 1 (POST) confirms auth and routing work on this URL."
echo "  Probes 2-4 (GET) consistently return 500 INTERNAL_ERROR with"
echo "  refreshed web-auth tokens (auth succeeded; handler then crashed)."
echo "================================================================"
```

### Captured probe matrix from 2026-05-08

```
Probe                                              | Status | Server message
--------------------------------------------------+--------+----------------------------------------------------------
1. POST users/discover, empty array (control)     | 400    | BAD_REQUEST: discover request contains an invalid user id
2. GET users/discover, no body                    | 500    | INTERNAL_ERROR (no reason)
3. GET users/discover, Accept: application/json   | 500    | INTERNAL_ERROR
4. GET users/discover, explicit empty body        | 500    | INTERNAL_ERROR
5. GET private/users/discover (control)           | 400    | endpoint not applicable in the database type 'privatedb'
6. GET shared/users/discover (control)            | 400    | endpoint not applicable in the database type 'sharedb'
7. GET users/discoverAll (fake path, control)     | 404    | could not find handler for endpoint 'GET discoverAll'
8. OPTIONS users/discover                         | 200    | Access-Control-Allow-Methods: GET, POST, PUT, DELETE
```

---

### 2. CloudKitJS canary — `AuthTokenIndexHTML+ScriptDiscoverTest.swift`

Original path on the branch: `Examples/MistDemo/Sources/MistDemoKit/Commands/AuthTokenIndexHTML+ScriptDiscoverTest.swift`. A "Test discoverAllUserIdentities" button on `mistdemo auth-token`'s sign-in page that calls `container.discoverAllUserIdentities()` directly via Apple's `cloudkit.js`. To use, also wire `scriptDiscoverTest` into the rendered HTML alongside `scriptDisplay` and add the matching DOM elements (`discover-test-section`, `discover-test-button`, `discover-test-output`).

```swift
//
//  AuthTokenIndexHTML+ScriptDiscoverTest.swift
//  MistDemo
//

#if canImport(Hummingbird)
  // swiftlint:disable line_length indentation_width
  extension AuthTokenIndexHTML {
    /// JS for the CloudKitJS canary that exercises
    /// `container.discoverAllUserIdentities()` from the authenticated browser
    /// session, documenting the live behavior of the REST endpoint that
    /// MistKit's `discoverAllUserIdentities()` is `@available(*, unavailable)`
    /// against. Tracks issue #28.
    internal static let scriptDiscoverTest: String = #"""

              // CloudKitJS canary — calls Apple's own discoverAllUserIdentities()
              // method so we can observe whether it succeeds in an authenticated
              // browser. If it rejects with INTERNAL_ERROR here, the bug is on
              // Apple's side, not in MistKit's REST shape.
              const discoverTestSection = document.getElementById('discover-test-section');
              const discoverTestButton = document.getElementById('discover-test-button');
              const discoverTestOutput = document.getElementById('discover-test-output');

              function revealDiscoverTestSection() {
                  if (discoverTestSection) {
                      discoverTestSection.style.display = 'block';
                  }
              }

              async function runDiscoverAllUserIdentitiesTest() {
                  if (!container) {
                      showStatus('CloudKit container not initialized; sign in first.', true);
                      return;
                  }
                  if (typeof container.discoverAllUserIdentities !== 'function') {
                      showStatus('container.discoverAllUserIdentities is not a function — CloudKitJS may have removed it.', true);
                      return;
                  }
                  discoverTestButton.disabled = true;
                  discoverTestOutput.style.display = 'block';
                  discoverTestOutput.className = 'discover-test-output';
                  const startedAt = new Date().toISOString();
                  discoverTestOutput.textContent = 'Calling container.discoverAllUserIdentities()...\n\nstarted: ' + startedAt + '\n';

                  try {
                      const response = await container.discoverAllUserIdentities();
                      const finishedAt = new Date().toISOString();
                      const users = response && response.users ? response.users : [];
                      const summary = {
                          status: 'resolved',
                          started: startedAt,
                          finished: finishedAt,
                          hasErrors: response ? response.hasErrors : null,
                          errors: response ? response.errors : null,
                          userCount: users.length,
                          users: users
                      };
                      discoverTestOutput.className = 'discover-test-output success';
                      discoverTestOutput.textContent = JSON.stringify(summary, null, 2);
                      console.log('discoverAllUserIdentities resolved:', response);
                  } catch (error) {
                      const finishedAt = new Date().toISOString();
                      const summary = {
                          status: 'rejected',
                          started: startedAt,
                          finished: finishedAt,
                          'CKError.uuid': error && error.uuid ? error.uuid : null,
                          'CKError.serverErrorCode': error && error.serverErrorCode ? error.serverErrorCode : null,
                          'CKError.ckErrorCode': error && error.ckErrorCode ? error.ckErrorCode : null,
                          'CKError.reason': error && error.reason ? error.reason : null,
                          'CKError.retryAfter': error && error.retryAfter ? error.retryAfter : null,
                          'CKError.message': error && error.message ? error.message : null,
                          'error.toString()': String(error),
                          'error.toJSON()': error && typeof error.toJSON === 'function' ? error.toJSON() : null
                      };
                      discoverTestOutput.className = 'discover-test-output error';
                      discoverTestOutput.textContent = JSON.stringify(summary, null, 2);
                      console.error('discoverAllUserIdentities rejected:', error);
                  } finally {
                      discoverTestButton.disabled = false;
                  }
              }

              if (discoverTestButton) {
                  discoverTestButton.addEventListener('click', runDiscoverAllUserIdentitiesTest);
              }
      """#
  }
// swiftlint:enable line_length indentation_width
#endif
```

### Captured CloudKitJS canary result from 2026-05-08 19:43:20 UTC

Same defect, observed from inside Apple's own library. ~331ms round-trip.

```json
{
  "status": "rejected",
  "started": "2026-05-08T19:43:19.896Z",
  "finished": "2026-05-08T19:43:20.227Z",
  "CKError.uuid": "26a433f0-d9df-483a-b086-9b35b527ce8a",
  "CKError.serverErrorCode": "INTERNAL_ERROR",
  "CKError.ckErrorCode": "INTERNAL_ERROR",
  "CKError.reason": "INTERNAL_ERROR",
  "error.toJSON()": {
    "uuid": "26a433f0-d9df-483a-b086-9b35b527ce8a",
    "serverErrorCode": "INTERNAL_ERROR",
    "reason": "INTERNAL_ERROR",
    "ckErrorCode": "INTERNAL_ERROR"
  }
}
```

---

### CloudKitJS disassembly excerpt

For posterity, the relevant slice of `cdn.apple-cloudkit.com/ck/2/cloudkit.js` that proves CloudKitJS sends the byte-for-byte equivalent of the failing REST GET:

```js
discoverAllUserIdentities: function() {
  var e = (new RequestBuilder())
    .setApiEntityName("users")
    .setApiActionName("discover")
    .setResponseClass(UserIdentitiesResponse);
  return this.publicCloudDatabase.sendRequest(e);
}
// builder default: this._method = isNullOrUndefined(c) ? "GET" : "POST"
// setPayload() flips to POST. discoverAllUserIdentities never calls setPayload.
```

The sibling `fetchCurrentUserIdentity` (which targets `users/caller` and works fine, including via MistKit's `fetchCaller()`) uses the same builder pattern — same `GET`, same public DB, same no-payload. The only behavioral difference is the `discover` action name routing to a crashing handler.

**Comment by leogdion** [comment]
Feedback filed at FB22754466

**Comment by leogdion** [comment]
Here's the openradar link:
https://openradar.appspot.com/radar?id=EhAKBVJhZGFyEICAgPrBmPsJ
